### ISSUE TITLE
change var to let (introduce block scope) (previously function scope) (code refactor)

### ISSUE THREAD

**Issue by MohitBansal321** [openseadragon/openseadragon#2299]
**var has a Function scope**
Most programming languages have block scope. Why? Because it is less error-prone.
You can declare variables inside blocks (if statements, for loops, and so on) without worrying about overwriting some previously declared variable 
In Javascript, it doesn’t matter how often you use the keyword “var”. You are pointing to the same variable if it’s the same name in the same function.
**let are not just new cool keywords, they also introduce block scope that allows us to write clean and less error-prone code.**

**Comment by pearcetm** [comment]
This is currently not allowed, but will (should?) become possible once IE11 support is dropped, which will allow ES6 features (like `let`, `const`) to finally be used in OpenSeadragon core code. See https://github.com/openseadragon/openseadragon/issues/2286 for discussion of enabling this change.

**Comment by msalsbery** [comment]
Indeed! We would have loved to do it years ago!

We care a lot about backword compatibility... we'll get there  ;)

**Comment by MohitBansal321** [comment]
so how can we should do this?

**Comment by msalsbery** [comment]
> so how can we should do this?  @msalsbery  @pearcetm 

We're not going to do it quite yet

**Comment by MohitBansal321** [comment]
> > so how can we should do this?  @msalsbery  @pearcetm
> 
> We're not going to do it quite yet

So whenever you start doing inform me i will do changes

**Comment by AndrewADev** [comment]
> This is currently not allowed, but will (should?) become possible once IE11 support is dropped, which will allow ES6 features (like `let`, `const`) to finally be used in OpenSeadragon core code. See #2286 for discussion of enabling this change.

In the case of `let` and `const`, there is [actually quite wide support even in IE11](https://stackoverflow.com/a/2356926) (the special cases are related to usage in `for` loops).

It's linked in the S/O post as well, but those two on Can I use:
- `let`: https://caniuse.com/?search=let
- `const`: https://caniuse.com/?search=const

So, from a tech perspective, IE 11 support doesn't need to be a blocker for (initial) adoption of `let` and `const` (style and project preferences aside).

**Comment by iangilman** [comment]
Okay, I think we're ready to officially drop IE11 (see #2286). Let's not merge any ES6 until after the 4.0.1 deploy (#2302), but we can start working on patches. That said, I don't know that we need to rush into making changes just for change's sake. 

In the case of `var`, there are over 2000 cases of it in the code, and each one needs to be evaluated by hand before changing (since var is semantically different from let, even if we don't try to differentiate between let and const). I don't see that we gain any benefit from making the switch en masse; I would probably just recommend using the new syntax for new and updated code. But maybe there's some benefit I'm not seeing?

**Comment by msalsbery** [comment]
> Okay, I think we're ready to officially drop IE11 (see #2286). Let's not merge any ES6 until after the 4.0.1 deploy (#2302), but we can start working on patches. That said, I don't know that we need to rush into making changes just for change's sake. 
> 
> 
> 
> In the case of `var`, there are over 2000 cases of it in the code, and each one needs to be evaluated by hand before changing (since var is semantically different from let, even if we don't try to differentiate between let and const). I don't see that we gain any benefit from making the switch en masse; I would probably just recommend using the new syntax for new and updated code. But maybe there's some benefit I'm not seeing? 

Agree!!

**Comment by MohitBansal321** [comment]
> Okay, I think we're ready to officially drop IE11 (see #2286). Let's not merge any ES6 until after the 4.0.1 deploy (#2302), but we can start working on patches. That said, I don't know that we need to rush into making changes just for change's sake.
> 
> In the case of `var`, there are over 2000 cases of it in the code, and each one needs to be evaluated by hand before changing (since var is semantically different from let, even if we don't try to differentiate between let and const). I don't see that we gain any benefit from making the switch en masse; I would probably just recommend using the new syntax for new and updated code. But maybe there's some benefit I'm not seeing?

I Agree but  let it block scope variable and var is function scope variable 
eg. 
```
var x = 10;
// Here x is 10
{
  let x = 2;
  // Here x is 2
}
// Here x is 10
````
another changes normal function to arrow function
Using const is safer than using var because a function expression is always a constant value.

**Comment by msalsbery** [comment]
> 
> 
> I Agree but  let it block scope variable and var is function scope variable 
> 
> eg. 
> 
> ```
> 
> var x = 10;
> 
> // Here x is 10
> 
> {
> 
>   let x = 2;
> 
>   // Here x is 2
> 
> }
> 
> // Here x is 10
> 
> ````
> 
> another changes normal function to arrow function
> 
> Using const is safer than using var because a function expression is always a constant value.

Definitely. From here on we can use the ECMAScript 2015 stuff but there's no benefit I'm aware of to changing the entire code base since, as @iangilman mentioned, every change has to be carefully looked at to make sure behavior doesn't change and there's thousands of them  ;)

**Comment by MohitBansal321** [comment]
> > I Agree but  let it block scope variable and var is function scope variable
> > eg.
> > ```
> > 
> > var x = 10;
> > 
> > // Here x is 10
> > 
> > {
> > 
> >   let x = 2;
> > 
> >   // Here x is 2
> > 
> > }
> > 
> > // Here x is 10
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > another changes normal function to arrow function
> > Using const is safer than using var because a function expression is always a constant value.
> 
> Definitely. From here on we can use the ECMAScript 2015 stuff but there's no benefit I'm aware of to changing the entire code base since, as @iangilman mentioned, every change has to be carefully looked at to make sure behavior doesn't change and there's thousands of them ;)

agree it is challenging to change 2830 var to let but one solution maybe we can do is 
file by file not entirely 
one other is normal function to arrow function

**Comment by msalsbery** [comment]
> 
> agree it is challenging to change 2830 var to let but one solution maybe we can do is 
> 
> file by file not entirely 
> 
> one other is normal function to arrow function 

The benefit/risk ratio is close to 0 unless someone comes up with a compelling benefit

**Comment by AndrewADev** [comment]
Not sure if this has been mentioned here, but for getting rid of `var` (and preferring other parts of ES 6 syntax/best practices - e.g. `const` over `let` when no reassignement), there [is an `eslint` rule](https://eslint.org/docs/latest/rules/no-var), many of which `eslint` will autofix if you pass it the `--fix` option:

![image](https://user-images.githubusercontent.com/2937540/221273947-b1d88f34-b6c1-4924-aea8-3d6324917e13.png)

Optionally, those could be combined with something like a pre-commit hook for limiting to staged changes.

**Comment by iangilman** [comment]
@AndrewADev I'd forgotten you can use ESLint to fix… that seems like a pretty safe way to make a bulk changes like this.

If we were to do anything like that, I think I would be in favor of doing it as a one-off thing, rather than including ESLint fix into our workflow.
